File: useless/parser/lexer.c
#include "headers/lexer.h"
#include "headers/expression.h"
#include "headers/new_string.h"
#include "headers/useless_util.h"
#include <stdio.h>
#include <stdlib.h>
int main(int argc, char *argv[]) {
  FILE *ptr;
  ptr = fopen(argv[1], "r");
  if (ptr == NULL) {
    return EXIT_FAILURE;
  }
  int BUFFER_SIZE = 4096;
  char buffer[BUFFER_SIZE];
  while (fgets(buffer, BUFFER_SIZE, ptr)) {
    StringList *rows = split(buffer, ',', -1);
    printf("row size(%d): \n", rows->size);
    if (rows->size < 3) {
      printf("skip line\n");
      continue;
    }
    strip_mark(rows->strings[1]);
    printf("input: %s: \n", rows->strings[1]->value);
    Tokens *tokens = lexer(rows->strings[1]->value);
    printf("tokens size(%d): \n", tokens->size);
    printTokens(tokens);
    printf("\niter end\n");
  }
  return EXIT_SUCCESS;
}
Tokens *lexer(char *str) {
  Tokens *tokens = array_new();
  int i = 0;
  while (*str) {
    if (!whiteList(*str)) {
      goto Next;
    }
    if (isWhiteSpace(*str)) {
      goto Next;
    }
    if (isOperator(tokens, *str, i)) {
      goto Next;
    }
    if (isParentheses(tokens, *str, i)) {
      goto Next;
    }
    if (consumeInteger(tokens, &str, &i)) {
      goto Next;
    }
    if (consumeFloat(tokens, &str, &i)) {
      goto Next;
    }
  Next:
    str++;
    i++;
  }
  return tokens;
}
int whiteList(char c) {
  if (c == '\x09' || c == '\x20') {
    return 1;
  }
  if (c >= '0' && c <= '9') {
    return 1;
  }
  if (c == '.' || c == ',') {
    return 1;
  }
  if (c == '(' || c == ')') {
    return 1;
  }
  if (c == '+' || c == '-' || c == '*' || c == '/') {
    return 1;
  }
  return 0;
}
int consumeWhiteSpace(Tokens *array, char **str, int *index) {
  if (!isWhiteSpace(**str)) {
    return 0;
  }
  Token token;
  int i = (*index)++;
  (*str)++;
  while (**str) {
    if (!isWhiteSpace(**str)) {
      goto Valid;
    }
    (*str)++;
    (*index)++;
  }
  return 1;
Valid:
  token.type = WHITESPACE;
  token.span.start = i;
  token.span.end = *index - 1;
  appendToken(array, &token);
  return 0;
}
int isWhiteSpace(char c) {
  if (c == SPACE || c == HORIZONTAL_TAB) {
    return 1;
  }
  return 0;
}
int isOperator(Tokens *tokens, char c, int index) {
  switch (c) {
  case Add:
  case Substract:
  case Multply:
  case Divide:
    setTokenChar(tokens, OPERATOR, c, index, index + 1);
    return 1;
  }
  return 0;
}
int isParentheses(Tokens *array, char c, int index) {
  if (c == '(' || c == ')') {
    setTokenChar(array, PARENTHESES, c, index, index + 1);
    return 1;
  }
  return 0;
}
int consumeInteger(Tokens *tokens, char **str, int *index) {
  if (!isNumber(**str)) {
    return 0;
  }
  char *startChar = (*str)++;
  int startIndex = (*index)++;
  for (; **str; (*str)++) {
    if (!isNumber(**str)) {
      if (**str == '.') {
        *str = startChar;
        *index = startIndex;
        return 0;
      }
      break;
    }
    (*index)++;
  }
  setToken(tokens, INTEGER, *str - *index, startIndex, *index);
  (*str)--;
  (*index)--;
  return 1;
}
//^-?\d+\.\d+$
int consumeFloat(Tokens *tokens, char **str, int *index) {
  if (!isNumber(**str)) {
    return 0;
  }
  char *startChar = (*str)++;
  int startIndex = (*index)++;
  for (; **str; (*str)++) {
    if (!isNumber(**str)) {
      if (**str != '.') {
        *str = startChar;
        *index = startIndex;
        return 0;
      }
      (*str)++;
      (*index)++;
      break;
    }
    (*index)++;
  }
  for (; **str; (*str)++) {
    if (!isNumber(**str)) {
      break;
    }
    (*index)++;
  }
  setToken(tokens, FLOAT, *str - *index, startIndex, *index);
  (*str)--;
  (*index)--;
  return 1;
}
int consumerNumber(Tokens *array, char **str, int *index) {
  if (!isNumber(**str)) {
    return 0;
  }
  if (**str == '0' && !isSeperator(*(*str + 1))) {
    setToken(array, INTEGER, *str, *index, ++(*index));
    (*str)++;
    return 0;
  }
  int i = *index;
  while (**str) {
    if (!isNumber(**str)) {
      if (**str == '.' || **str == ',') {
        (*str)++;
        (*index)++;
        if (isRightSideFloat(str, index)) {
          setToken(array, FLOAT, *str - *index, i, *index);
          return 0;
        }
        return 0;
      }
      setToken(array, INTEGER, *str - *index, i, *index);
      return 0;
    }
    (*str)++;
    (*index)++;
  }
  setToken(array, INTEGER, *str - *index, i, *index);
  return 1;
}
void setToken(Tokens *array, Type type, char *cstring, int start, int end) {
  String *string = substring(cstring, start, end);
  Span span = {start, end};
  Token token = {type, string, span};
  appendToken(array, &token);
}
void setTokenChar(Tokens *array, Type type, char c, int start, int end) {
  String *string = string_new();
  string_append_char(string, c);
  Span span = {start, end};
  Token token = {type, string, span};
  appendToken(array, &token);
}
//^(0|\d+)$
int isInteger(Tokens *array, char *str, int index) {}
//^(0.\d+|\d.\d+)$
int isRightSideFloat(char **str, int *index) {
  if (!isNumber(**str)) {
    return 0;
  }
  (*str)++;
  (*index)++;
  for (; *str; (*index)++) {
    if (!isNumber(**str)) {
      return 1;
    }
    (*str)++;
  }
  return 1;
}
int isSeperator(char c) {
  if (c == '.' || c == ',') {
    return 1;
  }
  return 0;
}
Tokens *array_new() {
  Tokens *ptr = (Tokens *)malloc(sizeof(Tokens));
  if (ptr == NULL) {
    return (Tokens *)NULL;
  }
  Token *tokens_ptr = (Token *)malloc(sizeof(Token) * 10);
  if (tokens_ptr == NULL) {
    return (Tokens *)NULL;
  }
  ptr->tokens = tokens_ptr;
  ptr->size = 0;
  ptr->capacity = 10;
  return ptr;
}
void appendToken(Tokens *array, Token *token) {
  if (array->size + 1 > array->capacity) {
    int capacity = roundCapacity(array->size + 1);
    Token *ptr = (Token *)realloc(array->tokens, sizeof(Token) * capacity);
    if (ptr == NULL) {
      return;
    }
    array->tokens = ptr;
    array->capacity = capacity;
  }
  memcpy(array->tokens + array->size, token, sizeof(Token));
  array->size++;
}
String *getToken(Type type) {
  String *token = string_new_value("<UNKNOWN>");
  switch (type) {
  case INTEGER:
    string_set_cstrings(token, "<INTEGER> -> ");
    break;
  case FLOAT:
    string_set_cstrings(token, "<FLOAT> -> ");
    break;
  case OPERATOR:
    string_set_cstrings(token, "<OPERATOR> -> ");
    break;
  case PARENTHESES:
    string_set_cstrings(token, "<PARENTHESES> -> ");
    break;
  case WHITESPACE:
    string_set_cstrings(token, "<WHITESPACE> -> ");
    break;
  }
  return token;
}
int getSpanSize(Span span) { return (span.end - span.start) + 1; }
void printTokens(Tokens *tokens) {
  for (int i = 0; i < tokens->size; i++) {
    Token token = tokens->tokens[i];
    printf("%s", getToken(token.type)->value);
    printf("<%s>\n", token.string->value);
    // printf("<%.*s>,", getSpanSize(token.span), str + token.span.start);
  }
  printf("\n");
}
int isTokenChar(Token token, char c) { return token.string->value[0] == c; }

File: zig/src/lexer.zig
pub const Lexer = @This();
const std = @import("std");
const Expr = @import("expression.zig");
pub const Type = enum(u4) {
    INTEGER,
    FLOAT,
    OPERATOR,
    PARENTHESES,
};
pub const Token = struct {
    type: Type,
    start: u32,
    end: u32,
};

const SPACE = @as(u8, 0x20);
const HORIZONTAL_TAB = @as(u8, 0x09);
pub fn lexer(arena: std.mem.Allocator, code: []const u8) ![]Token {
    var tokens = try std.ArrayList(Token).initCapacity(arena, 100);
    var i: u32 = 0;
    while (i < code.len) : (i += 1) {
        const char = code[i];
        if (!isInWhiteList(char)) continue;
        if (isWhiteSpace(char)) continue;
        if (isOperator(char)) {
            try tokens.append(arena, .{
                .type = .OPERATOR,
                .start = i,
                .end = i + 1,
            });
            continue;
        }
        if (isParentheses(char)) {
            try tokens.append(arena, .{
                .type = .PARENTHESES,
                .start = i,
                .end = i + 1,
            });
            continue;
        }
        if (try comsumeInteger(arena, &tokens, code, &i)) continue;
        if (try consumeFloat(arena, &tokens, code, &i)) continue;
    }
    return tokens.items;
}
pub fn isInWhiteList(char: u8) bool {
    if (isWhiteSpace(char)) return true;
    if (isNumber(char)) return true;
    if (isOperator(char)) return true;
    if (char == '.' or char == ',') return true;
    if (char == '(' or char == ')') return true;
    return false;
}
pub fn isWhiteSpace(char: u8) bool {
    return char == SPACE or char == HORIZONTAL_TAB;
}
pub fn isOperator(char: u8) bool {
    const value: Expr.Operator = @enumFromInt(char);
    switch (value) {
        .ADD,
        .SUBSTRACT,
        .MULTPLY,
        .DIVIDE,
        => return true,
        _ => return false,
    }
}
pub fn isParentheses(char: u8) bool {
    return char == '(' or char == ')';
}
pub fn comsumeInteger(arena: std.mem.Allocator, tokens: *std.ArrayList(Token), code: []const u8, index: *u32) !bool {
    if (!isNumber(code[index.*])) return false;
    const start: u32 = index.*;
    index.* += 1;
    while (index.* < code.len) : (index.* += 1) {
        const char = code[index.*];
        if (isNumber(char)) continue;
        if (char == '.') {
            index.* = start;
            return false;
        }
        break;
    }

    try tokens.append(arena, .{
        .type = .INTEGER,
        .start = start,
        .end = index.*,
    });
    index.* -= 1;
    return true;
}
pub fn isNumber(char: u8) bool {
    return char >= '0' and char <= '9';
}
pub fn consumeFloat(arena: std.mem.Allocator, tokens: *std.ArrayList(Token), code: []const u8, index: *u32) !bool {
    if (!isNumber(code[index.*])) return false;
    const start: u32 = index.*;
    index.* += 1;
    while (index.* < code.len) : (index.* += 1) {
        const char = code[index.*];
        if (isNumber(char)) continue;
        if (char == '.') break;
        index.* = start;
        return false;
    }
    index.* += 1;
    while (index.* < code.len) : (index.* += 1) {
        const char = code[index.*];
        if (!isNumber(char)) break;
    }
    try tokens.append(arena, .{
        .type = .FLOAT,
        .start = start,
        .end = index.*,
    });

    index.* -= 1;
    return true;
}
pub fn isTokenChar(token: Token, code: []const u8, char: u8) bool {
    return getValue(token, code)[0] == char;
}
pub fn printToken(writer: *std.Io.Writer, token: Token, code: []const u8) !void {
    try writer.print("<{s}> -> ", .{getTokenType(token)});
    try writer.print("<{s}>\n", .{getValue(token, code)});
    try writer.flush();
}
pub fn printTokens(writer: *std.Io.Writer, tokens: []Token, code: []const u8) !void {
    for (tokens) |token| {
        try writer.print("<{s}> -> ", .{getTokenType(token)});
        try writer.print("<{s}>\n", .{getValue(token, code)});
        try writer.flush();
    }
}
pub fn printTokenValues(writer: *std.Io.Writer, tokens: []Token, code: []const u8) !void {
    for (tokens) |token| {
        try writer.print("{s}", .{getValue(token, code)});
        try writer.flush();
    }
}
pub fn getValue(token: Token, code: []const u8) []const u8 {
    return code[token.start..token.end];
}
pub fn getTokenType(token: Token) []const u8 {
    return @tagName(token.type);
}
